regex-capi/src/lib.rs

```diff
@@ -0,0 +1,515 @@
+extern crate libc;
+extern crate regex;
+
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::fmt;
+use std::io::{self, Write};
+use std::ops::Deref;
+use std::panic::{self, AssertUnwindSafe};
+use std::process;
+use std::ptr;
+use std::slice;
+use std::str;
+
+use libc::{c_char, uint32_t};
+use regex::bytes;
+
+macro_rules! protect {
```

Comment:
This could theoretically incorporate more of the FFI set-up, e.g.

```rust
macro_rules! ffi_fn {
     (fn $name: ident($($arg: ident: $arg_ty: ty),*) -> $ret: ty $body: block) => {
          #[no_mangle]
          pub extern fn $name($($arg: $arg_ty),*) -> $ret { ... }
     }
}

ffi_fn! {
    fn rust_compile_must(pattern: *const c_char) -> *const Regex { ... }
}
```